### PR TITLE
Vector tiles

### PR DESCRIPTION
**Pull request by klinkow**

### CODE REVIEW

File: seedsource_core/django/seedsource/management/commands/create_vector_tiles.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+            if process.returncode == 0:
+                self.stdout.write(self.style.SUCCESS("Success\n"))
+                outputIndex.append({'name': name, 'type': 'vector', 'urlTemplate': f'http://localhost:3333/services/seedzones/{uid}' + "/tiles/{z}/{x}/{y}.png", 'zIndex': 1, 'displayed': False})
+            else:
+                errors.append(name)
+                self.stdout.write(self.style.ERROR("Error\n"))
+
+        self._write_out("Cleaning up temp files..")
+        os.remove(tiles_dir + "/temp/geojson")
```
Comment (nikmolnar) [inline seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: This should be done in a `try`/`finally`. Also, have a look at [tempfile](https://docs.python.org/3.6/library/tempfile.html).

```python
from tempfile import mkdtemp

...

tmp_dir = mkdtemp()
try:
    with open(os.path.join(tmp_dir, 'zones.json')) as f:
        f.write(sz.polygon.json)

    ...

finally:
    try:
        shutil.rmtree(tmp_dir)
    except OSError:
         pass  # Ignore errors removing temp dir
```

File: seedsource_core/django/seedsource/management/commands/create_vector_tiles.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        os.rmdir(tiles_dir + "/temp")
+
+        self._write_out("Creating index..")
+        with open(tiles_dir + "/index.js", "w") as f:
```
Comment (nikmolnar) [inline seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: How does this get used?

Comment (klinkow) [reply seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: It's a list of all that's getting served by mbtiles in the same format as layer objects found in the layer reducer. I copy it into my project config.js file and then in layers reducer I set it to my default state. Eventually I was thinking of having this write straight to the project config.js file.

**Comment by klinkow** [comment]
Changes made.

File: seedsource_core/django/seedsource/management/commands/create_vector_tiles.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import os
+from django.conf import settings
+import json
+
+
+class Command(BaseCommand):
+    help = 'Facilitates converting of vector data into vector tiles.'
+
+    def _write_out(self, output):
+        self.stdout.write('\033[0;33m' + output + '\033[0m')
```
[the remark this answers is not in the document]
Comment (nikmolnar) [reply seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: Nice!

File: seedsource_core/django/seedsource/management/commands/create_vector_tiles.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import shutil
+
+
+class Command(BaseCommand):
+    help = 'Facilitates converting of vector data into vector tiles.'
+
+    def _write_out(self, output):
+        self.stdout.write('\033[0;33m' + output + '\033[0m')
+
+    def _create_folder(self, directory):
```
Comment (nikmolnar) [inline seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: Given that this is only two lines and only used once, I think it makes more sense to simply inline the code. It's actually more code overall this way.

File: seedsource_core/django/seedsource/management/commands/create_vector_tiles.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            source = source.replace("/", "-").lower()
+            tmp_dir = mkdtemp()
+
+            self._write_out(f'Loading seedzones of source "{source}" ...')
+            geojson = {
+                'type': 'FeatureCollection',
+                'features': [json.loads(sz.polygon.geojson) for sz in zones]
+            }
+
+            try:
```
Comment (nikmolnar) [inline seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: The `try` block ideally should start right after `tmp_dir = mkdtemp()`. That way the temp dir always gets cleaned up, even if lines 35-39 happen to error.

File: seedsource_core/django/seedsource/management/commands/create_vector_tiles.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    def handle(self, *args, **options):
+        tiles_dir = os.path.join(settings.BASE_DIR, "tiles")
+        outputIndex = []
+        errors = []
+        sources = SeedZone.objects.values_list('source').distinct()
+
+        self._create_folder(os.path.join(tiles_dir, "seedzones"))
+
+        for source in sources:
+            source = str(source).strip("'(,)'")
```
Comment (nikmolnar) [inline seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: `source` should already be a string.

**Comment by nikmolnar** [comment]
I ran the `create_vector_tiles` command, but other than creating a `seedzones` folder and an `index.js`, nothing else seems to have happened. It returned almost immediately and didn't create any `.mbtiles` files.

Ideas?

File: seedsource_core/django/seedsource/management/commands/create_vector_tiles.py (hunk 6 above)
Comment (klinkow) [reply seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: unfortunately it's a tuple... I replaced the garbage above with source[0]

Comment (nikmolnar) [reply seedsource_core/django/seedsource/management/commands/create_vector_tiles.py]: Ah, ok. Actually just add `flat=True` to the `values_list` call.

```python
sources = SeedZone.objects.values_list('source', flat=True).distinct()
```

**Comment by klinkow** [comment]
Hmmm, it's working on my machine. It sounds like it's completely skipping tippecanoe. That's strange it doesn't give any additional output?

**Comment by nikmolnar** [comment]
See my edit above.

**Comment by klinkow** [comment]
O good. That's interesting though I would have thought tippecanoe would something like 'no files loaded'

**Comment by nikmolnar** [comment]
It was never trying to run tippecanoe because the loop would run 0 times.

**Comment by klinkow** [comment]
O right

**Comment by klinkow** [comment]
: )

**Pull request merged**